fix: enable full decimal to decimal support

himadripal opened the pull request:
Closes #375 

- enable decimal to decimal
-  remove hard coded  castoptions to pass to native execution
- fixed castTest to match arrow invalid argument error with spark's Number out of range error.

## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #.

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

## How are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

parthchandra commented on native/spark-expr/src/conversion_funcs/cast.rs:
```diff
@@ -872,6 +872,13 @@ fn cast_array(
 
+    let native_cast_options: CastOptions = CastOptions {
+        safe: !matches!(cast_options.eval_mode, EvalMode::Ansi), // take safe mode from cast_options passed
+        format_options: FormatOptions::new()
```
I think one can use a `default` value defined for `FormatOptions` here

himadripal replied:
In the default CAST_OPTIONS which is replaced by this native_cast_option had two these set to 
```rust 
static TIMESTAMP_FORMAT: Option<&str> = Some("%Y-%m-%d %H:%M:%S%.f");
            timestamp_format: TIMESTAMP_FORMAT,
            timestamp_tz_format: TIMESTAMP_FORMAT,
```
If we change it to `default`, I checked the FormatOptions.default() implementation does not set the `timestamp_format` and `timestamp_tz_format`
values. Hence kept it as it is defined in CAST_OPTIONS.

parthchandra replied:
Fair enough. (The format options are used only to make the cast of timestamp to string compatible with Spark, and are not needed anywhere else) but I guess it is a good idea to be consistent everywhere.

kazuyukitanimura commented on spark/src/main/scala/org/apache/comet/expressions/CometCast.scala:
```diff
@@ -70,13 +70,8 @@ object CometCast {
-          Compatible()
-        }
+      case (_: DecimalType, _: DecimalType) =>
+        Compatible()
```
Would you mind updating docs/source/user-guide/compatibility.md

kazuyukitanimura commented on spark/src/test/scala/org/apache/comet/CometCastSuite.scala:
```diff
@@ -913,8 +913,8 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
       .toDF("b")
-      .withColumn("a", col("b").cast(DecimalType(6, 2)))
-    checkSparkAnswer(df)
+      .withColumn("a", col("b").cast(DecimalType(38, 28)))
```
We can create the test case like
```
    val rowData = Seq(
      Row(BigDecimal("12345.6789")),
      Row(BigDecimal("9876.5432")),
      Row(BigDecimal("123.4567"))
    )
    val df = spark.createDataFrame(
      spark.sparkContext.parallelize(rowData),
      StructType(Seq(StructField("a", DataTypes.createDecimalType(10,4))))
    )
```
so that we can avoid to have `cast` in the input df

kazuyukitanimura commented on docs/source/user-guide/compatibility.md:
```diff
@@ -73,76 +73,77 @@ Spark.
-|-|-|-|
-| boolean | byte |  |
-| boolean | short |  |
+|-|---------|-|
```
nit: Just checking whether this change is due to the changes in the producing method
Should be automatically created by `make release`

kazuyukitanimura commented on spark/src/test/scala/org/apache/comet/CometCastSuite.scala:
```diff
@@ -1126,27 +1129,36 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
-                  sparkException.getMessage
-                    .replace(".WITH_SUGGESTION] ", "]")
-                    .startsWith(cometMessage))
-              } else if (CometSparkSessionExtensions.isSpark34Plus) {
-                // for Spark 3.4 we expect to reproduce the error message exactly
-                assert(cometMessage == sparkMessage)
+                  cometMessage.contains("cannot be represented as") || cometMessage.contains(
+                    "too large to store"))
               } else {
```
There are message modifications below per spark version
Would you mind update them instead of creating another if branch?

kazuyukitanimura reviewed (commented):
Mostly looks good, thank you @himadripal just minor comments

kazuyukitanimura commented on spark/src/test/scala/org/apache/comet/CometCastSuite.scala:
```diff
@@ -1126,27 +1129,33 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
-              } else if (CometSparkSessionExtensions.isSpark34Plus) {
-                // for Spark 3.4 we expect to reproduce the error message exactly
-                assert(cometMessage == sparkMessage)
+              // for comet decimal conversion throws ArrowError(string) from arrow - across spark versions the message dont match.
+              if (sparkMessage.contains("cannot be represented as")) {
+                cometMessage.contains("cannot be represented as") || cometMessage.contains(
+                  "too large to store")
               } else {
```
I think we still need to remove this new `if` block and update the test cases below.
This new block may still pass with cometMessage.contains("cannot be represented as") that seems to be an indication of Spark cast instead of native cast

himadripal replied:
when I removed the branch from the top, the test fails for double to decimal conversion with allow-incompatible flag, I think that is still using spark cast. Hence it needs to be there.

himadripal replied:
> Another way to remove the if block is to convert the error message to make it similar to Spark
This is the place that we are defining Spark error messages https://github.com/apache/datafusion-comet/blob/main/native/spark-expr/src/error.rs#L36
You can check how these are used

This one I checked - problem here is from native execution, we get back

himadripal replied:
> Or you can move the message check below and switch the expected messages based on the fromType/toType.

I'll try this.

himadripal replied:
> One or the other way. Otherwise we cannot confidently say that the test is passing due to the cannot be represented as message or the too large to store message.

I'm contemplating creating a different test/check function for decimal to decimal test.

himadripal replied:
@kazuyukitanimura  please check one more time.

himadripal commented on spark/src/main/scala/org/apache/comet/GenerateDocs.scala:
```diff
@@ -69,7 +69,8 @@ object GenerateDocs {
         for (fromType <- CometCast.supportedTypes) {
           for (toType <- CometCast.supportedTypes) {
-            if (Cast.canCast(fromType, toType) && fromType != toType) {
+            if (Cast.canCast(fromType, toType) && (fromType != toType || fromType.typeName
```
@andygrove  please check - I added this exception for decimal

andygrove replied:
This check looks good

andygrove commented on spark/src/test/scala/org/apache/comet/CometCastSuite.scala:
```diff
@@ -1126,27 +1129,33 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
+                  assert(cometMessage == sparkMessage)
                 } else {
-                  assert(cometMessageModified == sparkMessage)
+                  // for Spark 3.3 we just need to strip the prefix from the Comet message
```
note that we plan on dropping 3.x support soon so can simplify the code here once we do that

andygrove commented:
Apologies for the late review @himadripal. I checked out the branch locally and ran some additional manual testing and could not find any issues. I left a comment about a missing assertion. Could you also rebase against main since this has been pending for a while?

It would be great to get this into the 0.7.0 release next week if possible.

himadripal commented:
> I left a comment about a missing assertion

@andygrove  I did not see this comment. Although I reverted previous changes related to spark3.3 assertion and added a check for specific messages for decimal to decimal conversion.

andygrove commented on spark/src/test/scala/org/apache/comet/CometCastSuite.scala:
```diff
@@ -1210,27 +1213,36 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
-                    .startsWith(cometMessage))
-              } else if (CometSparkSessionExtensions.isSpark34Plus) {
-                // for Spark 3.4 we expect to reproduce the error message exactly
-                assert(cometMessage == sparkMessage)
+                  cometMessage.contains("too large to store"),
+                  sparkMessage.contains("cannot be represented as"))
```
This doesn't seem correct. The second argument to `assert` is usually a message that gets displayed if the assertion fails.

himadripal replied:
It was meant to be `==` in between instead of `,`.

kazuyukitanimura replied:
Right now, `false == false` will pass the test as well. I think this should be two lines
```
assert(cometMessage.contains("too large to store"))
assert(sparkMessage.contains("cannot be represented as"))
```

himadripal replied:
my though was `false == false` is valid use case, in case some other exception happened in spark other than "can not be represented", then comet should also throw some other exception, not "too large to store". I can probably make it explicit by putting the assert inside if.
```scala
    if (sparkMessage.contains("cannot be represented as") {
          assert(cometMessage.contains("too large to store"))
    }
```

WDYT?

kazuyukitanimura reviewed (approved):
LGTM (pending ci)

andygrove reviewed (approved):
Thanks @himadripal
